File: 06.design-patterns/gift-selection/java/src/main/java/gift/GiftSelector.java
package gift;

public class GiftSelector {
    private GiftSelector() {
    }

    public static String selectGiftFor(Child child) {
        if (child.behavior() == Behavior.NAUGHTY) {
            return null;
        } else if (child.behavior() == Behavior.NORMAL) {
            return child.giftRequests()
                    .reversed()
                    .stream()
                    .filter(GiftRequest::isFeasible)
                    .map(GiftRequest::giftName)
                    .findFirst()
                    .orElse(null);
        } else {
            return child.giftRequests().stream()
                    .filter(GiftRequest::isFeasible)
                    .map(GiftRequest::giftName)
                    .findFirst()
                    .orElse(null);
        }
    }
}

File: 06.design-patterns/gift-selection/ts/src/giftSelector.ts
import {Child} from "./child";

export function selectGiftFor(child: Child): string | null {
    if (child.behavior === "naughty") {
        return null;
    } else if (child.behavior === "normal") {
        return [...child.giftRequest]
            .reverse()
            .filter(gift => gift.isFeasible)
            .map(gift => gift.giftName)[0] ?? null;
    } else if (child.behavior === "nice") {
        return child.giftRequest
            .filter(gift => gift.isFeasible)
            .map(gift => gift.giftName)[0] ?? null;
    }
}